### PR TITLE
72 hierarchical faceting investigation

### LINKED ISSUE #72: Investigation for community model
Can we change ```hasCollection: [noid, noid, noid], belongsToCommunity: [noid, noid, noid]``` to some kind of single property nested structure such as

```memberOf: [[community-noid, collection-noid], [community-noid, collection-noid], ...]```

to eliminate ordering dependencies and issues with multiple non-unique ids appearing in properties? While retaining the ability to facet on collections or communities?

Can we get acceptable performance while still removing names (and thus further ordering dependencies) from Works by moving the name lookup to solr, either by fetching individual records of collections or fetching (and potentially caching) one master noid->name mapping in solr (while avoiding the performance penalty of engaging in any reifycation of Fedora objects to get these names)? (this has the nice side-effect that renaming a collection would no longer imply updating every member object's collection properties)

edit: with recent discussion of the unspecified ordering of metadata fields in Fedora, this is especially critical

### PR DESCRIPTION
**Pull request by mbarnett**
Adds support for Hierarchical faceting

Works in concert with solrizer changes to support ancestor_ and descendant_path types. Closes #72

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -54,7 +62,10 @@ group :development do
 
   # Spring speeds up development by keeping your application running in the background. Read more: https://github.com/rails/spring
   gem 'spring'
+  gem 'pry', '~> 0.10.4'
+  gem 'better_errors'
 
+  gem 'rb-readline', platforms: :ruby, install_if: os_is(/darwin/)
```
Comment (murny) [inline Gemfile:72]: Why not something more simple like:
```ruby
  gem 'rb-readline', platforms: :ruby, require: RUBY_PLATFORM =~ /darwin/i ? 'rb-readline' : false
```
(is platforms needed even?)

(RUBY_PLATFORM contains your OS version for example mine is RUBY_PLATFORM ==> "x86_64-linux")
Then you don't need the helper os method above?

File: Gemfile [hunk 2]
```diff
@@ -54,7 +62,10 @@ group :development do
 
   # Spring speeds up development by keeping your application running in the background. Read more: https://github.com/rails/spring
   gem 'spring'
+  gem 'pry', '~> 0.10.4'
```
Comment (murny) [inline Gemfile:66]: ¯\_(ツ)_/¯ Can we use `pry-rails` instead? (It is pry but also it replaces your console to use pry instead of irb which is a better console)

File: Gemfile (hunk 1 above)
Comment (murny) [reply Gemfile:72]: Are we even using rb-readline here? Also it seems to be a gem that any OS can use? Why limit it to OSX?

https://github.com/ConnorAtherton/rb-readline

Comment (mbarnett) [reply Gemfile:72]: cruby 2.3.1 is broken on OS X (for at least some people, myself included) with linker errors to unresolved references in readline.dylib. AFAIK mruby, rubinius, jruby etc isn't.

tl;dr I need it to make ruby work, folks on linux don't. I don't think it makes sense to force it onto anyone else, including possibly anyone who wants to pull this project down and try it on non-cruby on OS X. It makes sense in my mind to clearly document that this is a specific platform/os combination install. Removing platforms: and just using your regex would install it more broadly than it needs to be.

As for the method, same argument. It better semantically documents what's going on here than an inline regex.

File: Gemfile (hunk 2 above)
Comment (mbarnett) [reply Gemfile:66]: Do you like pry-rails? I don't have a strong preference either way but I guess the argument for pry is, it's there when you call it and not there otherwise, whereas pry-rails kind of forces itself onto eveyrone. I'm actually not wild about not being able to _not_ be in pry when I don't want to be, personally.

File: Gemfile (hunk 1 above)
Comment (murny) [reply Gemfile:72]: Does this deserve to be in the application codebase then If it's just to fix your machine so you can get ruby running? Why not install this locally? `gem install rb-readline`? Maybe I'm confused about what this is all for? ¯_(ツ)_/¯

Comment (mbarnett) [reply Gemfile:72]: Not putting it in the gem file means there's defacto undocumented but required gems that some members of the team need to know to install to get work done. That seems like a bad idea to me. 

It's going to get annoying if I have to manually install that every time I start up a new gemset but if this is a huge deal to you I can pull this method out (and put it back in in a couple of weeks when we add clamav and it doesn't build for me...)

File: Gemfile (hunk 2 above)
Comment (murny) [reply Gemfile:66]: There the exact same? Except pry-rails overwrites your rails c to use pry instead of irb? Which is a better console I would argue? Either way I don't care to much either honestly. Just recommending it. Can always get a pry console by using `bundle exec pry -r ./config/environment`

Comment (mbarnett) [reply Gemfile:66]: That's what I mean. I don't like that it overwrites 'rails c'. I'd rather only be in pry when I explicitly call into it, personally

File: Gemfile (hunk 1 above)
Comment (murny) [reply Gemfile:72]: I'm not understanding maybe? But its your ruby installation that's broken and requiring this config correct? If so why you need this in the app? If you use a higher ruby version is this needed? If you use rvm instead of chruby maybe? Looks like you can just `brew install readline` instead from quickly googling this https://github.com/guard/guard/wiki/Add-Readline-support-to-Ruby-on-Mac-OS-X?

 I mean if you need it to run the app then fine add it. But please document this? Cause there is nothing about what this is for or what this does. Other then it is required for Mac OSX users which isn't exactly true? My Macbook at home can run ruby/rails code without this in the gemfile? And I have never seen popular open source rails apps requiring this either in their gemfiles (and most people dev in Mac/OSX)

File: Gemfile (hunk 2 above)
Comment (murny) [reply Gemfile:66]: Fair enough. If no one else wants pry as their default console then leave it how it is. (You can always opt out of pry as your console by setting an ENV VAR (DISABLE_PRY_RAILS=1)  if you were to add pry-rails instead)

Comment (mbarnett) [reply Gemfile:66]: Ah I didn't realize that. I'm fine either way, in that case.

File: Gemfile (hunk 1 above)
Comment (mbarnett) [reply Gemfile:72]: Yeah, the homebrewing a more recent readline didn't work for me (which seems to be the case for at least a few other people as well: https://github.com/rails/rails/issues/26658)

I'm leaving it in place but I'll add a note explaining it & to reevaluate down the road.
